jenkins/pipelines/cd/atom-jobs/products/tiflash/experiment.groovy

```diff
@@ -0,0 +1,457 @@
+                git tag -f ${Version} ${GitHash}
+                git branch -D refs/tags/${Version} || true
+                git checkout -b refs/tags/${Version}
+            """
+            sh getBuildCMD(disableCache)
+            sh """
+                cd output
+                tar --exclude=tiflash.tar.gz -czvf tiflash.tar.gz tiflash
+                sha256sum tiflash.tar.gz | cut -d ' ' -f 1 >tiflash.tar.gz.sha256
+            """
+        }
+        stage("upload"){
+        dir("output"){
+            sh "curl -F $BinPath=@tiflash.tar.gz ${FILE_SERVER_URL}/upload"
+            sh "curl -F ${BinPath}.sha256=@tiflash.tar.gz.sha256 ${FILE_SERVER_URL}/upload"
+        }
+        }
+        stage("upload proxy cache"){
+            if(need_update_proxy && proxy_cache_path){
+                sh """
+                tar -czvf tiflash_proxy.tar.gz -C output/tiflash libtiflash_proxy.${dylib_postfix()}
+                curl -F $proxy_cache_path=@tiflash_proxy.tar.gz ${FILE_SERVER_URL}/upload
+                """
+            }else{
+                echo "skip"
+            }
+        }
+}
+
+
+
+def buildBin={
+    def skip = false
+    stage("check"){
+        def binExist = sh(script: "curl -I ${getBinDownloadURL()}|grep \"200 OK\"", returnStatus: true)
+        if (binExist  == 0) {
+            skip = true
+        }
+    }
+    if (skip) {
+        echo "Binary exists, skip build"
+        return
+    }
+    dir('tiflash'){
+        doBuild()
+    }
+}
+
+
+def buildDocker={
+    sh 'printenv HUB_PSW | docker login -u $HUB_USR --password-stdin hub.pingcap.net'
+    sh """
+        curl --fail --retry 3 -o Dockerfile https://raw.githubusercontent.com/PingCAP-QE/artifacts/main/dockerfiles/tiflash.Dockerfile
+        curl --fail --retry 3 -o tiflash.tar.gz ${getBinDownloadURL()}
+        tar -xzvf tiflash.tar.gz
+        rm -f tiflash.tar.gz
+        docker build -t ${DockerImage}-$ARCH .
+        docker push ${DockerImage}-$ARCH
+    """ 
+}
+
+
+pipeline{
+    parameters {
+        string(name: 'GitHash', description: 'the git tag or commit', defaultValue:'7bd02a86e08f5077c69299ac048a700fcb68507f')
+        string(name: 'Version', description: 'important, the Version for cli --Version and profile choosing, eg. v6.5.0', defaultValue:'v7.3.0')
+        choice(name: 'Edition', choices : ["community", "enterprise"])
+        string(name: 'PathForLinuxAmd64', defaultValue: '', description: 'build path linux amd64')
+        string(name: 'PathForLinuxArm64', defaultValue: '', description: 'build path linux arm64')
+        string(name: 'PathForDarwinAmd64', defaultValue: '', description: 'build path darwin amd64')
+        string(name: 'PathForDarwinArm64', defaultValue: '', description: 'build path darwin arm64')
+        booleanParam(name: 'CleanBuild', description:"disable all caches")
+        string(name: 'DockerImage', description: 'docker image path', defaultValue: '')
```

Comment:
Consider whether to add a new parameter "ForceRebuild": Sometimes it may be necessary to recompile, even if the commit hash has already been compiled and uploaded to the fileserver.

Reply:
CleanBuild will skip any cache